actors/miner/src/lib.rs

```diff
@@ -1950,6 +1958,306 @@ impl Actor {
+                sealed_cid: sector.sealed_cid,
+                deprecated_deal_ids: vec![],
+                expiration: sector.expiration,
+                activation: activation_epoch,
+                deal_weight: DealWeight::zero(),
+                verified_deal_weight: DealWeight::zero(),
+                initial_pledge: initial_pledge.clone(),
+                expected_day_reward: day_reward.clone(),
+                expected_storage_pledge: storage_pledge.clone(),
+                power_base_epoch: activation_epoch,
+                replaced_day_reward: TokenAmount::zero(),
+                sector_key_cid: None,
+                flags: SectorOnChainInfoFlags::SIMPLE_QA_POWER,
+            })
+            .collect::<Vec<SectorOnChainInfo>>();
+        let sectors_len = sectors_to_add.len();
+
+        let total_pledge = BigInt::from(sectors_len) * initial_pledge;
+        let mut needs_cron = false;
+
+        rt.transaction(|state: &mut State, rt| {
+            let current_balance = rt.current_balance();
+            let available_balance = state
+                .get_unlocked_balance(&current_balance)
+                .with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+                    "failed to calculate unlocked balance"
+                })?;
+            if available_balance < total_pledge {
+                return Err(actor_error!(
+                    insufficient_funds,
+                    "insufficient funds for aggregate initial pledge requirement {}, available: {}",
+                    total_pledge,
+                    available_balance
+                ));
+            }
+            needs_cron = !state.deadline_cron_active;
+            state.deadline_cron_active = true;
+
+            state
+                .put_sectors(store, sectors_to_add.clone())
+                .with_context_code(ExitCode::USR_ILLEGAL_STATE, || "failed to put new sectors")?;
+
+            sectors_to_add.sort_by_key(|info| info.sector_number);
+            state
+                .assign_sectors_to_deadline(
+                    policy,
+                    store,
+                    rt.curr_epoch(),
+                    sectors_to_add,
+                    info.window_post_partition_sectors,
+                    info.sector_size,
+                    params.proving_deadline,
+                )
+                .with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+                    format!("failed to assign new sectors to deadline {}", params.proving_deadline)
+                })?;
+
+            state
+                .add_initial_pledge(&total_pledge)
+                .with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+                    "failed to add initial pledgs"
+                })?;
+
+            state
+                .check_balance_invariants(&rt.current_balance())
+                .map_err(balance_invariants_broken)?;
+
+            Ok(())
+        })?;
+
+        pay_aggregate_seal_proof_fee(rt, sectors_len)?;
```

Comment:
somewhere around here we need `sector-activated` events per sector, see comment in the test [below](https://github.com/filecoin-project/builtin-actors/pull/1537/files#r1640853964)